nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/ExecutorDao.java

```diff
@@ -147,25 +145,18 @@ public void updateActiveTimestamp() {
   public void recoverWorkflowInstancesFromDeadNodes() {
     List<InstanceInfo> instances = jdbc.query(
         "select id, state from nflow_workflow where executor_id in (select id from nflow_executor where "
-            + executorGroupCondition + " and id <> ? and expires < current_timestamp)", new RowMapper<InstanceInfo>() {
+            + executorGroupCondition + " and id <> " + getExecutorId() + " and expires < current_timestamp)",
+        new RowMapper<InstanceInfo>() {
           @Override
           public InstanceInfo mapRow(ResultSet rs, int rowNum) throws SQLException {
             InstanceInfo instance = new InstanceInfo();
-            instance.id = rs.getInt(1);
-            instance.state = rs.getString(2);
+            instance.id = rs.getInt("id");
+            instance.state = rs.getString("state");
             return instance;
           }
-        }, getExecutorId());
+        });
     for (InstanceInfo instance : instances) {
-      // TODO: move this block to WorkflowInstanceDao and add transaction
-      int updated = jdbc.update("update nflow_workflow set executor_id = null where id = ? and executor_id in (select id from nflow_executor where " + executorGroupCondition + " and id <> ? and expires < current_timestamp)",
-          instance.id, getExecutorId());
-      if (updated > 0) {
-        WorkflowInstanceAction action = new WorkflowInstanceAction.Builder().setExecutionStart(now()).setExecutionEnd(now())
-            .setExecutorId(getExecutorId()).setType(recovery).setState(instance.state).setStateText("Recovered")
-            .setWorkflowInstanceId(instance.id).build();
-        workflowInstanceDao.insertWorkflowInstanceAction(action);
-      }
+      workflowInstanceDao.recoverWorkflowInstance(instance);
```

Comment:
Should we still create the action in ExexcutorDao? And pass (long instanceId, WorkflowInstanceAction action) as arguments to instanceDao. Then instanceDao would not have to know about the InstanceInfo object at all.